vault/barrier_aes_gcm.go

```diff
@@ -559,9 +580,14 @@ func (b *AESGCMBarrier) CheckUpgrade(ctx context.Context) (bool, uint32, error)
 
 // ActiveKeyInfo is used to inform details about the active key
 func (b *AESGCMBarrier) ActiveKeyInfo() (*KeyInfo, error) {
+	if atomic.LoadUint32(b.sealed) == 1 {
+		return nil, ErrBarrierSealed
+	}
+
 	b.l.RLock()
 	defer b.l.RUnlock()
-	if b.sealed {
+
+	if atomic.LoadUint32(b.sealed) == 1 {
```

Comment:
Yeah, the reason for this pattern is that we can be sure that values like b.keyring won't change if we have the lock, but we *can't* be sure that b.sealed won't. So using the atomic is a fast way to check whether we should even bother acquiring the lock, and in some functions it's completely sufficient (like List, Delete, and Sealed), but otherwise after grabbing the lock we need to make sure that Seal didn't run first.

Reply:
It basically slightly disadvantages Put and Get (in that they now need to do an atomic load instead of a straight bool read) while providing a real advantage for Delete and List, but the other changes I made in Put and Get allow us to give up the lock sooner, so I'm thinking the tradeoff is probably fine?

Reply:
Also I think for Put and Get the time is likely to be dominated by actual decryption/encryption vs an atomic load of a 32-bit uint, whereas avoiding a lock grab in delete/list is quite a decent speedup relatively speaking.